src/worker/mod.rs

```diff
@@ -139,39 +140,45 @@ impl Worker {
                 max_cached_workflows: config.max_cached_workflows,
             }
         };
-        let (wau_tx, wau_rx) = unbounded_channel();
+        let (pan_tx, pan_rx) = watch::channel(true);
+        let (wftd_tx, wftd_rx) = watch::channel(false);
         let (shut_tx, shut_rx) = watch::channel(false);
         Self {
             server_gateway: sg.clone(),
             sticky_name: sticky_queue_name,
             wf_task_poll_buffer: wft_poller,
-            wft_manager: WorkflowTaskManager::new(wau_tx, cache_policy),
+            wft_manager: WorkflowTaskManager::new(pan_tx, cache_policy),
             at_task_mgr: act_poller
                 .map(|ap| WorkerActivityTasks::new(config.max_outstanding_activities, ap, sg)),
             workflows_semaphore: Semaphore::new(config.max_outstanding_workflow_tasks),
             config,
             shutdown_requested: shut_rx,
             shutdown_sender: shut_tx,
-            workflow_activations_update: Mutex::new(wau_rx),
+            wfts_drained: wftd_rx,
+            wfts_drained_sender: wftd_tx,
+            pending_activations_notification_receiver: Mutex::new(pan_rx),
         }
     }
 
     /// Tell the worker to begin the shutdown process. Can be used before [shutdown_complete] if
     /// polling should be ceased before it is possible to consume the worker instance.
     pub(crate) async fn notify_shutdown(&self) {
         let _ = self.shutdown_sender.send(true);
-        self.wf_task_poll_buffer.notify_shutdown();
-        if let Some(b) = self.at_task_mgr.as_ref() {
-            b.notify_shutdown();
+        if let Some(atm) = self.at_task_mgr.as_ref() {
+            atm.notify_shutdown();
         }
-        loop {
-            // Wait until all outstanding workflow tasks have been completed before shutting down
-            if self.outstanding_workflow_tasks() != 0 {
-                let mut waulock = self.workflow_activations_update.lock().await;
-                let _ = waulock.recv().await;
-                continue;
-            }
-            break;
+        self.wf_task_poll_buffer.notify_shutdown();
+        // Notify in case shutdown was requested while there were no more outstanding WFTs.
+        // This is required because the only other place where we notify wfts_drained is on
+        // activation completion and activation polling checks for wfts_drained.
+        self.maybe_notify_wtfs_drained();
+        // wait until all outstanding workflow tasks have been completed before shutting down
+        if !*self.wfts_drained.borrow() {
+            self.wfts_drained
+                .clone()
+                .changed()
```

Comment:
I realized while I was looking at some of this that this should probably be moved to `shutdown_complete`, since it waits all this time. That does slightly change some of the existing semantics though where workers will end up gone instead of returning `Shutdown`. 

I don't suuuper care if we move it, or if we maybe just update the docstring for this function, but we should do one of the two.